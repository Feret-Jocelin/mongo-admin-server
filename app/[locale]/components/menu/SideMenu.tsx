'use client';
import { useState } from 'react';
import { useTranslations } from 'next-intl';
import { useRouter } from 'next/navigation';

import styles from './menu.module.scss';
import '@/shared/styles/main.scss';

import GenericButton from '@components/ui/button/GenericButton';
import SelectInput from '@components/ui/inputs/select/SelectInput';
import CollectionList from '@components/ui/list/collection/CollectionList';
import LanguageModal from '@components/modal/language/LanguageModal';

import { DatabaseType } from '@/domain/entities/database-types';
import { logout } from '@/domain/usecases/auth-slice';

import { useSelector, useDispatch } from '@/store/store';
import {
  selectDatabaseSelected,
  selectDatabases,
  setDatabaseSelected,
} from '@/domain/usecases/database-slice';
import {
  selectCollectionSelected,
  selectCollectionByDatabase,
  setCollectionSelected,
  selectLoadingCollection,
} from '@/domain/usecases/collection-slice';
import { setTheme, selectTheme, selectLanguage } from '@/domain/usecases/setting-slice';
import { setIsLogged } from '@/domain/usecases/auth-slice';
import { database } from '@/infrastructure';

const SideMenu = () => {
  const dispatch = useDispatch();
  const t = useTranslations();
  const router = useRouter();
  const [openLanguageModal, setOpenLanguageModal] = useState(false);

  const language = useSelector(selectLanguage);
  const databaseSelected = useSelector(selectDatabaseSelected);
  const databases = useSelector(selectDatabases).map(
    (database: DatabaseType) => {
      return { value: database.name, label: database.name };
    }
  );

  const collectionSelected = useSelector(selectCollectionSelected);
  const collections = useSelector(selectCollectionByDatabase).map(
    (collection) => collection.collectionName
  );
  const loadingCollection = useSelector(selectLoadingCollection);

  const theme = useSelector(selectTheme);

  const handleChangeTheme = () => {
    theme === 'light' ? dispatch(setTheme('dark')) : dispatch(setTheme('light'));
  };

  const handleDatabaseChange = (database: string) => {
    router.push(`/${language}/dashboard/${database}`);
    dispatch(setDatabaseSelected(database));
  };

  const handleCollectionChange = (collection: string) => {
    dispatch(setCollectionSelected(collection));
    router.push(`/${language}/dashboard/${databaseSelected}/${collection}`);
  };

  const handleShowListUsers = () => {
    router.push(`/${language}/dashboard/admin/user`)
    console.log("user");
  }
  const handleLogout = () => {
    dispatch(logout());
    dispatch(setIsLogged(false));
    router.push(`/`);
  };

  return (
    <>
      <aside className={styles.aside}>
        <section className={styles.header}>
          {/* Logo */}

          <SelectInput
            label={t('menuSideBar.database')}
            value={databaseSelected}
            options={databases}
            onChange={(e) => handleDatabaseChange(e.target.value)}
          />

          {(databaseSelected && !loadingCollection) &&  (
            <section className={styles.contents}>
              <h5>{t('menuSideBar.collection')}</h5>
              <div className={`${styles.collections} scrollable`}>
                <CollectionList
                  collections={collections}
                  collectionSelected={collectionSelected}
                  onClick={(collection) =>
                    handleCollectionChange(collection)
                  }
                />
              </div>
            </section>
          )}
        </section>

        <section className={styles.footer}>
          <hr className={styles.divider} />
          <GenericButton
            icon_name="flag"
            padding="0 20px"
            transparent
            onClick={() => setOpenLanguageModal(true)}
          >
            {t('menuSideBar.language')}
          </GenericButton>

          <GenericButton
            icon_name={theme === 'light' ? 'moon' : 'sun'}
            padding="0 20px"
            transparent
            onClick={handleChangeTheme}
          >
            {theme === 'light' ? t('theme.dark') : t('theme.light')}
          </GenericButton>

          <GenericButton
            icon_name="gear"
            padding="0 20px"
            transparent
            onClick={() => console.log('Click settings')}
          >
            {t('menuSideBar.setting')}
          </GenericButton>
          <GenericButton
            icon_name="gear"
            padding="0 20px"
            transparent
            onClick={handleShowListUsers}
          >
            {t('menuSideBar.user')}
          </GenericButton>

          <GenericButton
            icon_name="logout"
            padding="0 20px"
            transparent
            variant='danger'
            onClick={handleLogout}
          >
            {t('menuSideBar.logout')}
          </GenericButton>
        </section>
      </aside>

      <LanguageModal
        open={openLanguageModal}
        onClose={() => setOpenLanguageModal(false)}
      />
    </>
  );
};

export default SideMenu;
