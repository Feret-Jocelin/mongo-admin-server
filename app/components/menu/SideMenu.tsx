'use client';
import { useEffect, useState } from "react";

import styles from './menu.module.scss';

import GenericButton from "@components/ui/button/GenericButton";
import SelectInput from "@components/ui/inputs/select/SelectInput";

import getDatabase from "@/infrastructure/databaseGateway";
import { DatabaseType } from "@/domain/entities/database-types";

const SideMenu = () => {
  const [databaseSelected, setDatabaseSelected] = useState('');
  const [listDatabase, setListDatabase] = useState<string[]>([]);

  useEffect(() => {
    fetchAllDatabase();
  }, []);

  const fetchAllDatabase = async () => {
    const database = await getDatabase();
    setListDatabase(database.map((db: DatabaseType) => db.name));
  }

  return (
    <aside className={styles.aside}>
      <section className={styles.header}>
        {/* Logo */}

        <SelectInput
          label="Base de données"
          value={databaseSelected}
          options={listDatabase}
          onChange={(e) => setDatabaseSelected(e.target.value)}
        />

        {databaseSelected && (
          <section className={styles.collections}>
            <h5>Collections</h5>
            {/* Collections */}
          </section>
        )}
      </section>

      <section className={styles.footer}>
        <hr className={styles.divider} />
        <GenericButton
          icon_name="caretUpDown"
          width="100%"
          border="none"
          color="#FFFFFF"
          background="transparent"
          padding="0 20px"
          onClick={() => console.log('Click language')}
        >
          Français
        </GenericButton>

        <GenericButton
          icon_name="sun"
          width="100%"
          border="none"
          color="#FFFFFF"
          background="transparent"
          padding="0 20px"
          onClick={() => console.log('Click theme')}
        >
          Langue
        </GenericButton>

        <GenericButton
          icon_name="gear"
          width="100%"
          border="none"
          color="#FFFFFF"
          background="transparent"
          padding="0 20px"
          onClick={() => console.log('Click setting')}
        >
          Paramètres
        </GenericButton>

        <GenericButton
          icon_name="logout"
          width="100%"
          border="none"
          color="#EB5757"
          background="transparent"
          padding="0 20px"
          onClick={() => console.log('Click logout')}
        >
          Déconexion
        </GenericButton>
      </section>
    </aside>
  );
}

export default SideMenu;