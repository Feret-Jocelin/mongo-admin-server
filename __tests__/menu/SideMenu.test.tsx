/* eslint-disable no-undef */
import React from 'react';
import { render } from '@testing-library/react';
import { NextIntlClientProvider } from 'next-intl';
import { store } from '@/store/store';
import { Provider } from 'react-redux';
import SideMenu from '@/app/[locale]/components/menu/SideMenu';

jest.mock('next-intl/client', () => ({
  usePathname() {
    return '';
  },
  useRouter() {
    return {
      push: () => jest.fn(),
      replace: () => jest.fn(),
    };
  },
}));
jest.mock('next/navigation', () => ({
  useRouter() {
    return {
      push: () => jest.fn(),
      replace: () => jest.fn(),
    };
  },
}));
const locale = 'fr';
const message = require(`../../shared/locales/${locale}.json`)

describe("SideMenu", () => {
  it('renders', () => {
    const handleClose = jest.fn();
    const { container } = render(
      <Provider store={store}>
        <NextIntlClientProvider locale={locale} messages={message}>
          <SideMenu />
        </NextIntlClientProvider>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  })
})