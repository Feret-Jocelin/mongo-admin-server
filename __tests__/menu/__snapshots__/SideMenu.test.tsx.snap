// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideMenu renders 1`] = `
<div>
  <aside
    class="aside"
  >
    <section
      class="header"
    >
      <div
        class="container"
      >
        <label>
          Base de donnÃ©es
        </label>
        <select
          class="select "
        >
          <option
            disabled=""
            value=""
          >
            Choisir
          </option>
        </select>
        <div
          class="iconContainer"
        >
          <div
            class="undefined "
          />
        </div>
      </div>
    </section>
    <section
      class="footer"
    >
      <hr
        class="divider"
      />
      <button
        class="button full transparent"
        style="padding: 0px 20px;"
        type="button"
      >
        <div
          class="icon "
        />
        Langue
      </button>
      <button
        class="button full transparent"
        style="padding: 0px 20px;"
        type="button"
      >
        <div
          class="icon "
        />
        Sombre
      </button>
      <button
        class="button full transparent"
        style="padding: 0px 20px;"
        type="button"
      >
        <div
          class="icon "
        />
        ParamÃ¨tres
      </button>
      <button
        class="button full transparent danger"
        style="padding: 0px 20px;"
        type="button"
      >
        <div
          class="icon "
        />
        DÃ©connexion
      </button>
    </section>
  </aside>
  <dialog
    class="modal modal--closing"
  >
    <section
      class="modal__header"
    >
      <div
        class="modal__header--icon"
      >
        <div
          class="undefined pointer-cursor"
        />
      </div>
      <h2>
        Langue
      </h2>
    </section>
    <section
      class="modal__container"
    >
      <ul
        class="language-list"
      >
        <li
          class="language-container selected"
        >
          ðŸ‡«ðŸ‡· FranÃ§ais
        </li>
        <li
          class="language-container "
        >
          ðŸ‡¬ðŸ‡§ Anglais
        </li>
        <li
          class="language-container "
        >
          ðŸ‡ªðŸ‡¸ Espagnol
        </li>
      </ul>
    </section>
  </dialog>
</div>
`;
